Riot/Modules/Settings/SignOut/SignOutFlowPresenter.swift

```diff
@@ -0,0 +1,162 @@
+    /// The presenter encountered an error and has stopped.
+    func signOutFlowPresenter(_ presenter: SignOutFlowPresenter, didFailWith error: Error)
+}
+
+/// This class provides a reusable component to present the sign out flow
+/// for the current session, including the initial prompt, and any follow-up
+/// key-backup setup that is necessary for the user.
+@objcMembers class SignOutFlowPresenter: NSObject {
+    private let session: MXSession
+    private let presentingViewController: UIViewController
+    
+    private var signOutAlertPresenter = SignOutAlertPresenter()
+    
+    weak var delegate: SignOutFlowPresenterDelegate?
+    
+    init(session: MXSession, presentingViewController: UIViewController) {
+        self.session = session
+        self.presentingViewController = presentingViewController
+        
+        super.init()
+        
+        signOutAlertPresenter.delegate = self
+    }
+    
+    /// Starts the flow without a specific source view. On iPad any popups
+    /// will show from the presenting view controller itself.
+    func start() {
+        start(sourceView: presentingViewController.view)
+    }
+    
+    /// Starts the flow, presenting any popups on iPad from the specified view.
+    func start(sourceView: UIView?) {
+        guard let keyBackup = session.crypto?.backup else { return }
+        
+        signOutAlertPresenter.present(for: keyBackup.state,
+                                      areThereKeysToBackup: keyBackup.hasKeysToBackup,
+                                      from: presentingViewController,
+                                      sourceView: sourceView ?? presentingViewController.view,
+                                      animated: true)
+    }
+    
+    // MARK: - SecureBackupSetupCoordinatorBridgePresenter
+    
+    private var secureBackupSetupCoordinatorBridgePresenter: SecureBackupSetupCoordinatorBridgePresenter?
+    private var crossSigningSetupCoordinatorBridgePresenter: CrossSigningSetupCoordinatorBridgePresenter?
+
+    private func showSecureBackupSetupFromSignOutFlow() {
+        if canSetupSecureBackup {
+            setupSecureBackup2()
+        } else {
+            // Set up cross-signing first
+            setupCrossSigning(title: VectorL10n.secureKeyBackupSetupIntroTitle,
+                              message: VectorL10n.securitySettingsUserPasswordDescription) { [weak self] result in
+                guard let self = self else { return }
+                switch result {
+                case .success(let isCompleted):
+                    if isCompleted {
+                        self.setupSecureBackup2()
+                    }
+                case .failure(let error):
+                    self.delegate?.signOutFlowPresenter(self, didFailWith: error)
+                }
+            }
+        }
+    }
+    
+    private var canSetupSecureBackup: Bool {
+        return session.vc_canSetupSecureBackup()
+    }
+    
+    private func setupSecureBackup2() {
```

Comment:
I imagine this code is just copy-pasted from before to not have regressions? It seems though that we could easily get rid of the "2" in the name to make the naming less confusing

Reply:
Good point, I'd assumed it was a v2 backup or something, but its definitely meaningless having it here.